### PR TITLE
chore: remove duplicated code and fix minor bugs

### PR DESCRIPTION
**Pull request by dejanzele**
# Pull Request Template

## Description

Please include a summary of the changes and the related issue (if any). Describe your changes in detail to help reviewers understand your contribution.

- **What is the purpose of this PR?**
- Cleanup of duplicated logic and minor bugfixes
- **What was changed?**
- Nothing feature-wise
- **Why was it changed?**
- Improve stability and reusability
- **Does this address any existing issues or enhancement requests?**
- No

Fixes # (issue)

## Type of change

Please select the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code Style Update (formatting, renaming)
- [x] Refactor (code changes that do not fix a bug or add a feature)
- [ ] Documentation Update
- [ ] Other (please describe):

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions to reproduce these tests. List any relevant details for your test configuration.

- **Test Configuration**:
    - Kubernetes Version:
    - Helm Version:
    - OS:

- **Test Steps**:
    1. Step 1
    2. Step 2
    3. ...

## Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: internal/controller/install/executor_controller.go [hunk 1]
```diff
@@ -76,17 +76,14 @@ type ExecutorReconciler struct {
...
-	logger.Info("Fetching Executor object from cache")
+	logger.Info("Reconciling object")
+
 	var executor installv1alpha1.Executor
-	if err := r.Client.Get(ctx, req.NamespacedName, &executor); err != nil {
-		if k8serrors.IsNotFound(err) {
-			logger.Info("Executor not found in cache, ending reconcile")
-			return ctrl.Result{}, nil
-		}
+	if miss, err := getObjectFromCache(ctx, r.Client, &executor, req.NamespacedName, logger); err != nil || miss {
```
Comment (nikola-jokic) [inline internal/controller/install/executor_controller.go:86]: How about this:
```go
if err := r.Get(ctx, req.NamespacedName, &executor); err != nil {
	return ctrl.Result{}, client.IgnoreNotFound(err)
}
```

**Review by nikola-jokic — COMMENTED** [review]
Left few comments :relaxed: LMKWYT

File: internal/controller/install/armadaserver_controller.go [hunk 2]
```diff
@@ -63,68 +63,35 @@ type ArmadaServerReconciler struct {
...
-				return ctrl.Result{}, err
-			}
-		}
-
-		// Stop reconciliation as the item is being deleted
-		return ctrl.Result{}, nil
+	cleanupF := func(ctx context.Context) error {
+		return r.deleteExternalResources(ctx, components, logger)
+	}
+	finish, err := checkAndHandleResourceDeletion(ctx, r.Client, &server, operatorFinalizer, cleanupF, logger)
```
[the remark this answers is not in the document]
Comment (dejanzele) [reply internal/controller/install/armadaserver_controller.go]: So this is a snippet which is duplicated in 8 contollers with around 99% similarity.

I'd rather have it in a single function which is well-tested and reused in all controllers.

I refactored it a bit so the function seems smaller and a bit cleaner, maybe `checkAndHandleResourceDeletion` can be renamed a bit better but I don't have a better idea currently.

File: internal/controller/install/common_helpers.go [hunk 3]
```diff
@@ -468,3 +475,113 @@ func addGoMemLimit(env []corev1.EnvVar, resources corev1.ResourceRequirements) [
...
+	switch iv.Kind() {
+	case reflect.Ptr, reflect.Slice, reflect.Map, reflect.Func, reflect.Interface:
+		return iv.IsNil()
+	default:
+		return false
+	}
+}
+
+// getObjectFromCache will get the object from the Kubernetes cache and return if it is missing or an error.
+func getObjectFromCache(
```
[the remark this answers is not in the document]
Comment (dejanzele) [reply internal/controller/install/common_helpers.go]: renamed

File: internal/controller/install/executor_controller.go (hunk 1 above)
Comment (dejanzele) [reply internal/controller/install/executor_controller.go]: Reason why I extracted it is to log behaviour in a uniform way and to be reused in all controllers as I expect the same thing to happen in each controller.

Reason for this cleanups is that similar logic was implemented multiple times in each controller and that led to many code duplication.

File: internal/controller/install/common_helpers.go [hunk 4]
```diff
@@ -468,3 +475,113 @@ func addGoMemLimit(env []corev1.EnvVar, resources corev1.ResourceRequirements) [
...
+
+		// Stop reconciliation as the item is being deleted
+		return true, nil
+	}
+	// The object is not being deleted, continue reconciliation
+	return false, nil
+}
+
+// upsertObject will create or update the object with the mutateFn if the resource is not nil.
+func upsertObject(
```
[the remark this answers is not in the document]
Comment (dejanzele) [reply internal/controller/install/common_helpers.go]: This is a function which does identical things in quite a few places and controllers.
I wanted it to be defined once, logged in exact same way (as previously there were nuances in logging which didn't make sense).
For the logger, I do like the structured logger, but for this specific log, it didn't have a lot of clarity.

**Review by dave-gantenbein — APPROVED** [review]
Nice refactor - ship it!

**Pull request merged**